% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={DADA2 analysis of 16S rRNA gene amplicon sequencing reads},
  pdfauthor={Jianshu Zhao},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi

\title{DADA2 analysis of 16S rRNA gene amplicon sequencing reads}
\author{Jianshu Zhao}
\date{2021-04-30}

\begin{document}
\maketitle

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{knitr}\SpecialCharTok{::}\NormalTok{opts\_chunk}\SpecialCharTok{$}\FunctionTok{set}\NormalTok{(}\AttributeTok{echo =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{knitr}\SpecialCharTok{::}\NormalTok{opts\_chunk}\SpecialCharTok{$}\FunctionTok{set}\NormalTok{(}\AttributeTok{fig.width =} \DecValTok{10}\NormalTok{)}
\NormalTok{knitr}\SpecialCharTok{::}\NormalTok{opts\_chunk}\SpecialCharTok{$}\FunctionTok{set}\NormalTok{(}\AttributeTok{tidy.opts=}\FunctionTok{list}\NormalTok{(}\AttributeTok{width.cutoff=}\DecValTok{60}\NormalTok{),}\AttributeTok{tidy=}\ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{introduction}{%
\subsection{Introduction}\label{introduction}}

Implementing DADA2 pipeline for resolving sequence variants from 16S
rRNA gene amplicon \textbf{\emph{paired-end}} sequencing reads, adopting
the tutorial from \url{https://benjjneb.github.io/dada2/tutorial.html}
and \url{https://benjjneb.github.io/dada2/bigdata_paired.html} with
minor adjustments. This report captures all the workflow steps necessary
to reproduce the analysis.

\hypertarget{load-r-packages}{%
\subsection{Load R packages:}\label{load-r-packages}}

\begin{verbatim}
## 
## Attaching package: 'RcppParallel'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:Rcpp':
## 
##     LdFlags
\end{verbatim}

\begin{verbatim}
## [1] '1.18.0'
\end{verbatim}

\begin{verbatim}
## 4
\end{verbatim}

\begin{verbatim}
## $BAND_SIZE
## [1] 16
## 
## $DETECT_SINGLETONS
## [1] FALSE
## 
## $GAP_PENALTY
## [1] -8
## 
## $GAPLESS
## [1] TRUE
## 
## $GREEDY
## [1] TRUE
## 
## $HOMOPOLYMER_GAP_PENALTY
## NULL
## 
## $KDIST_CUTOFF
## [1] 0.42
## 
## $MATCH
## [1] 5
## 
## $MAX_CLUST
## [1] 0
## 
## $MAX_CONSIST
## [1] 10
## 
## $MIN_ABUNDANCE
## [1] 1
## 
## $MIN_FOLD
## [1] 1
## 
## $MIN_HAMMING
## [1] 1
## 
## $MISMATCH
## [1] -4
## 
## $OMEGA_A
## [1] 1e-40
## 
## $OMEGA_C
## [1] 1e-40
## 
## $OMEGA_P
## [1] 1e-04
## 
## $PSEUDO_ABUNDANCE
## [1] Inf
## 
## $PSEUDO_PREVALENCE
## [1] 2
## 
## $SSE
## [1] 2
## 
## $USE_KMERS
## [1] TRUE
## 
## $USE_QUALS
## [1] TRUE
## 
## $VECTORIZED_ALIGNMENT
## [1] TRUE
\end{verbatim}

\begin{verbatim}
## /Users/jianshuzhao/Github/dada2_wrapper/scripts
\end{verbatim}

\hypertarget{get-list-of-input-fastq-files.}{%
\subsection{Get list of input fastq
files.}\label{get-list-of-input-fastq-files.}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Variable \textquotesingle{}input.path\textquotesingle{} containing path to input fastq files}
\CommentTok{\# directory is inherited from wrapper script dada2\_cli.r.}


\NormalTok{input.file.list }\OtherTok{\textless{}{-}} \FunctionTok{grep}\NormalTok{(}\StringTok{"*fastq"}\NormalTok{, }\FunctionTok{list.files}\NormalTok{(input.path), }\AttributeTok{value =}\NormalTok{ T)}
\CommentTok{\# input.path \textless{}{-} normalizePath(\textquotesingle{}input/\textquotesingle{})}

\CommentTok{\# List of input files}

\CommentTok{\# Sort ensures forward/reverse reads are in same order}
\NormalTok{fnFs }\OtherTok{\textless{}{-}} \FunctionTok{sort}\NormalTok{(}\FunctionTok{grep}\NormalTok{(}\StringTok{"\_R1.*}\SpecialCharTok{\textbackslash{}\textbackslash{}}\StringTok{.fastq"}\NormalTok{, }\FunctionTok{list.files}\NormalTok{(input.path), }\AttributeTok{value =}\NormalTok{ T))}
\NormalTok{fnRs }\OtherTok{\textless{}{-}} \FunctionTok{sort}\NormalTok{(}\FunctionTok{grep}\NormalTok{(}\StringTok{"\_R2.*}\SpecialCharTok{\textbackslash{}\textbackslash{}}\StringTok{.fastq"}\NormalTok{, }\FunctionTok{list.files}\NormalTok{(input.path), }\AttributeTok{value =}\NormalTok{ T))}

\CommentTok{\# Extract sample names, allowing variable filenames; e.g.}
\CommentTok{\# *\_R1[\_001].fastq[.gz]}
\NormalTok{sample.names }\OtherTok{\textless{}{-}} \FunctionTok{gsub}\NormalTok{(}\StringTok{"\_R1.*}\SpecialCharTok{\textbackslash{}\textbackslash{}}\StringTok{.fastq(}\SpecialCharTok{\textbackslash{}\textbackslash{}}\StringTok{.gz)?"}\NormalTok{, }\StringTok{""}\NormalTok{, fnFs, }\AttributeTok{perl =}\NormalTok{ T)}
\NormalTok{sample.namesR }\OtherTok{\textless{}{-}} \FunctionTok{gsub}\NormalTok{(}\StringTok{"\_R2.*}\SpecialCharTok{\textbackslash{}\textbackslash{}}\StringTok{.fastq(}\SpecialCharTok{\textbackslash{}\textbackslash{}}\StringTok{.gz)?"}\NormalTok{, }\StringTok{""}\NormalTok{, fnRs, }\AttributeTok{perl =}\NormalTok{ T)}
\ControlFlowTok{if}\NormalTok{ (}\SpecialCharTok{!}\FunctionTok{identical}\NormalTok{(sample.names, sample.namesR)) }\FunctionTok{stop}\NormalTok{(}\StringTok{"Forward and reverse files do not match."}\NormalTok{)}

\CommentTok{\# Specify the full path to the fnFs and fnRs}
\NormalTok{fnFs }\OtherTok{\textless{}{-}} \FunctionTok{file.path}\NormalTok{(input.path, fnFs)}
\NormalTok{fnRs }\OtherTok{\textless{}{-}} \FunctionTok{file.path}\NormalTok{(input.path, fnRs)}
\end{Highlighting}
\end{Shaded}

\hypertarget{generate-quality-plots-for-fwd-and-rev-reads-and-store-in-read_qc-folder.}{%
\subsection{Generate quality plots for FWD and REV reads and store in
Read\_QC
folder.}\label{generate-quality-plots-for-fwd-and-rev-reads-and-store-in-read_qc-folder.}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Create output folder }\AlertTok{NOTE}\CommentTok{: variable \textquotesingle{}output.dir\textquotesingle{} containing}
\CommentTok{\# name of output folder is inherited from wrapper script}
\CommentTok{\# dada2\_cli.r.}
\NormalTok{cwd }\OtherTok{\textless{}{-}} \FunctionTok{getwd}\NormalTok{()}

\NormalTok{readQC.folder }\OtherTok{\textless{}{-}} \FunctionTok{file.path}\NormalTok{(cwd, output.dir, }\StringTok{"Read\_QC"}\NormalTok{)}
\FunctionTok{ifelse}\NormalTok{(}\SpecialCharTok{!}\FunctionTok{dir.exists}\NormalTok{(readQC.folder), }\FunctionTok{dir.create}\NormalTok{(readQC.folder, }
    \AttributeTok{recursive =} \ConstantTok{TRUE}\NormalTok{), }\ConstantTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] TRUE
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Generate plots and save to folder in multi{-}page pdf}

\CommentTok{\# Forward reads}
\NormalTok{fwd.qc.plots.list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\FunctionTok{length}\NormalTok{(fnFs)) \{}
\NormalTok{    fwd.qc.plots.list[[i]] }\OtherTok{\textless{}{-}} \FunctionTok{plotQualityProfile}\NormalTok{(fnFs[i])}
    \FunctionTok{rm}\NormalTok{(i)}
\NormalTok{\}}
\DocumentationTok{\#\# plot the quality plot of forward reads of first sample}
\DocumentationTok{\#\# fwd.qc.plots.list[[1]] Save to file}
\NormalTok{mF }\OtherTok{\textless{}{-}} \FunctionTok{marrangeGrob}\NormalTok{(fwd.qc.plots.list, }\AttributeTok{ncol =} \DecValTok{2}\NormalTok{, }\AttributeTok{nrow =} \FunctionTok{length}\NormalTok{(fnFs)}\SpecialCharTok{/}\DecValTok{2} \SpecialCharTok{+} 
    \DecValTok{1}\NormalTok{, }\AttributeTok{top =} \ConstantTok{NULL}\NormalTok{)}
\FunctionTok{ggsave}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(readQC.folder, }\StringTok{"/FWD\_read\_plot.pdf"}\NormalTok{), mF, }\AttributeTok{width =} \DecValTok{7}\NormalTok{, }
    \AttributeTok{height =} \FloatTok{2.5} \SpecialCharTok{*}\NormalTok{ (}\FunctionTok{length}\NormalTok{(fnFs)}\SpecialCharTok{/}\DecValTok{2} \SpecialCharTok{+} \DecValTok{1}\NormalTok{), }\AttributeTok{device =} \StringTok{"pdf"}\NormalTok{, }\AttributeTok{limitsize =} \ConstantTok{FALSE}\NormalTok{)}
\FunctionTok{plot\_grid}\NormalTok{(fwd.qc.plots.list[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{ncol =} \DecValTok{2}\NormalTok{, }\AttributeTok{labels =} \StringTok{"auto"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{output_paired/16Sreport_dada2_pair2021-04-30_files/figure-latex/read_quality_plots-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{rm}\NormalTok{(fwd.qc.plots.list)}


\CommentTok{\# Reverse reads}
\NormalTok{rev.qc.plots.list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\FunctionTok{length}\NormalTok{(fnRs)) \{}
\NormalTok{    rev.qc.plots.list[[i]] }\OtherTok{\textless{}{-}} \FunctionTok{plotQualityProfile}\NormalTok{(fnRs[i])}
    \FunctionTok{rm}\NormalTok{(i)}
\NormalTok{\}}
\DocumentationTok{\#\# plot the quality plot of forward reads of first sample}
\DocumentationTok{\#\# rev.qc.plots.list[[1]]}
\NormalTok{mR }\OtherTok{\textless{}{-}} \FunctionTok{marrangeGrob}\NormalTok{(rev.qc.plots.list, }\AttributeTok{ncol =} \DecValTok{2}\NormalTok{, }\AttributeTok{nrow =} \FunctionTok{length}\NormalTok{(fnRs)}\SpecialCharTok{/}\DecValTok{2} \SpecialCharTok{+} 
    \DecValTok{1}\NormalTok{, }\AttributeTok{top =} \ConstantTok{NULL}\NormalTok{)}
\FunctionTok{ggsave}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(readQC.folder, }\StringTok{"/REV\_read\_plot.pdf"}\NormalTok{), mR, }\AttributeTok{width =} \DecValTok{7}\NormalTok{, }
    \AttributeTok{height =} \FloatTok{2.5} \SpecialCharTok{*}\NormalTok{ (}\FunctionTok{length}\NormalTok{(fnRs)}\SpecialCharTok{/}\DecValTok{2} \SpecialCharTok{+} \DecValTok{1}\NormalTok{), }\AttributeTok{device =} \StringTok{"pdf"}\NormalTok{, }\AttributeTok{limitsize =} \ConstantTok{FALSE}\NormalTok{)}
\FunctionTok{plot\_grid}\NormalTok{(rev.qc.plots.list[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{ncol =} \DecValTok{2}\NormalTok{, }\AttributeTok{labels =} \StringTok{"auto"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{output_paired/16Sreport_dada2_pair2021-04-30_files/figure-latex/read_quality_plots-2.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{rm}\NormalTok{(rev.qc.plots.list)}
\end{Highlighting}
\end{Shaded}

\hypertarget{trim-and-filter-reads.}{%
\subsection{Trim and filter reads.}\label{trim-and-filter-reads.}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Create filtered\_input/ subdirectory for storing filtered}
\CommentTok{\# fastq reads}
\NormalTok{filt\_path }\OtherTok{\textless{}{-}} \FunctionTok{file.path}\NormalTok{(cwd, output.dir, }\StringTok{"filtered\_input"}\NormalTok{)}
\FunctionTok{ifelse}\NormalTok{(}\SpecialCharTok{!}\FunctionTok{dir.exists}\NormalTok{(filt\_path), }\FunctionTok{dir.create}\NormalTok{(filt\_path, }\AttributeTok{recursive =} \ConstantTok{TRUE}\NormalTok{), }
    \ConstantTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] TRUE
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Define filenames for filtered input files}
\NormalTok{filtFs }\OtherTok{\textless{}{-}} \FunctionTok{file.path}\NormalTok{(filt\_path, }\FunctionTok{paste0}\NormalTok{(sample.names, }\StringTok{"\_F\_filt.fastq.gz"}\NormalTok{))}
\NormalTok{filtRs }\OtherTok{\textless{}{-}} \FunctionTok{file.path}\NormalTok{(filt\_path, }\FunctionTok{paste0}\NormalTok{(sample.names, }\StringTok{"\_R\_filt.fastq.gz"}\NormalTok{))}

\NormalTok{fnFs.noprimer }\OtherTok{\textless{}{-}} \FunctionTok{file.path}\NormalTok{(filt\_path, }\FunctionTok{paste0}\NormalTok{(sample.names, }\StringTok{"noprimer\_F\_filt.fastq.gz"}\NormalTok{))}
\NormalTok{fnRs.noprimer }\OtherTok{\textless{}{-}} \FunctionTok{file.path}\NormalTok{(filt\_path, }\FunctionTok{paste0}\NormalTok{(sample.names, }\StringTok{"noprimer\_R\_filt.fastq.gz"}\NormalTok{))}

\CommentTok{\# remove primers}
\ControlFlowTok{if}\NormalTok{ (}\SpecialCharTok{!}\FunctionTok{is.null}\NormalTok{(Fwd\_pr) }\SpecialCharTok{\&} \SpecialCharTok{!}\FunctionTok{is.null}\NormalTok{(Rvs\_pr)) \{}
    \FunctionTok{removePrimers}\NormalTok{(fnFs, fnFs.noprimer, }\AttributeTok{primer.fwd =}\NormalTok{ Fwd\_pr, }\AttributeTok{trim.fwd =} \ConstantTok{TRUE}\NormalTok{, }
        \AttributeTok{orient =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{verbose =} \ConstantTok{TRUE}\NormalTok{)}
    \FunctionTok{removePrimers}\NormalTok{(fnRs, fnRs.noprimer, }\AttributeTok{primer.fwd =}\NormalTok{ Rvs\_pr, }\AttributeTok{trim.fwd =} \ConstantTok{TRUE}\NormalTok{, }
        \AttributeTok{orient =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{verbose =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{\} }\ControlFlowTok{else}\NormalTok{ \{}
\NormalTok{    fnFs.noprimer }\OtherTok{\textless{}{-}}\NormalTok{ fnFs}
\NormalTok{    fnRs.noprimer }\OtherTok{\textless{}{-}}\NormalTok{ fnRs}
\NormalTok{\}}

\CommentTok{\# Filter the forward and reverse reads: Note that: 1. Reads}
\CommentTok{\# are both truncated and then filtered using the maxEE}
\CommentTok{\# expected errors algorighm from UPARSE.  2. Reverse reads}
\CommentTok{\# are truncated to shorter lengths than forward since they}
\CommentTok{\# are much lower quality.  3. \_Both\_ reads must pass for the}
\CommentTok{\# read pair to be output.  4. Output files are compressed by}
\CommentTok{\# default.}

\NormalTok{rd.counts }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{filterAndTrim}\NormalTok{(fnFs.noprimer, filtFs, }
\NormalTok{    fnRs.noprimer, filtRs, }\AttributeTok{truncLen =} \FunctionTok{c}\NormalTok{(}\DecValTok{200}\NormalTok{, }\DecValTok{190}\NormalTok{), }\AttributeTok{maxN =} \DecValTok{0}\NormalTok{, }
    \AttributeTok{maxEE =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{2}\NormalTok{), }\AttributeTok{truncQ =} \DecValTok{10}\NormalTok{, }\AttributeTok{rm.phix =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{compress =} \ConstantTok{TRUE}\NormalTok{, }
    \AttributeTok{multithread =}\NormalTok{ threads, }\AttributeTok{matchIDs =} \ConstantTok{TRUE}\NormalTok{))}
\CommentTok{\# Table of before/after read counts}
\NormalTok{rd.counts}\SpecialCharTok{$}\NormalTok{ratio }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(rd.counts}\SpecialCharTok{$}\NormalTok{reads.out}\SpecialCharTok{/}\NormalTok{rd.counts}\SpecialCharTok{$}\NormalTok{reads.in, }
    \AttributeTok{digits =} \DecValTok{2}\NormalTok{)}
\NormalTok{rd.counts}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                              reads.in reads.out ratio
## Orwoll_BI0023_BI_R1.fastq.gz    62724     53296  0.85
## Orwoll_BI0056_BI_R1.fastq.gz    55342     47545  0.86
## Orwoll_BI0131_BI_R1.fastq.gz    55144     46797  0.85
## Orwoll_BI0153_BI_R1.fastq.gz    44610     39510  0.89
## Orwoll_BI0215_BI_R1.fastq.gz    48227     40256  0.83
## Orwoll_BI0353_BI_R1.fastq.gz    54271     49323  0.91
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Write rd.counts table to file in readQC.folder}
\FunctionTok{write.table}\NormalTok{(rd.counts, }\FunctionTok{paste0}\NormalTok{(readQC.folder, }\StringTok{"/Read\_counts\_after\_filtering.tsv"}\NormalTok{), }
    \AttributeTok{sep =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{, }\AttributeTok{quote =}\NormalTok{ F, }\AttributeTok{eol =} \StringTok{"}\SpecialCharTok{\textbackslash{}n}\StringTok{"}\NormalTok{, }\AttributeTok{col.names =} \ConstantTok{NA}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{learn-the-error-rates.}{%
\subsection{Learn the error rates.}\label{learn-the-error-rates.}}

The DADA2 algorithm depends on a parametric error model (err) and every
amplicon dataset has a different set of error rates. The learnErrors
method learns the error model from the data, by alternating estimation
of the error rates and inference of sample composition until they
converge on a jointly consistent solution. As in many optimization
problems, the algorithm must begin with an initial guess, for which the
maximum possible error rates in this data are used (the error rates if
only the most abundant sequence is correct and all the rest are errors).

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{100}\NormalTok{)}
\CommentTok{\# Filtered forward reads}
\NormalTok{errF }\OtherTok{\textless{}{-}} \FunctionTok{learnErrors}\NormalTok{(filtFs, }\AttributeTok{nread =} \FloatTok{1e+06}\NormalTok{, }\AttributeTok{multithread =}\NormalTok{ threads, }
    \AttributeTok{randomize =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in learnErrors(filtFs, nread = 1e+06, multithread = threads, randomize
## = TRUE): The nreads parameter is DEPRECATED. Please update your code with the
## nbases parameter.
\end{verbatim}

\begin{verbatim}
## 55345400 total bases in 276727 reads from 6 samples will be used for learning the error rates.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Filtered reverse reads}
\NormalTok{errR }\OtherTok{\textless{}{-}} \FunctionTok{learnErrors}\NormalTok{(filtRs, }\AttributeTok{nread =} \FloatTok{1e+06}\NormalTok{, }\AttributeTok{multithread =}\NormalTok{ threads, }
    \AttributeTok{randomize =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in learnErrors(filtRs, nread = 1e+06, multithread = threads, randomize
## = TRUE): The nreads parameter is DEPRECATED. Please update your code with the
## nbases parameter.
\end{verbatim}

\begin{verbatim}
## 52578130 total bases in 276727 reads from 6 samples will be used for learning the error rates.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Visualize the estimated error rates Forward}
\FunctionTok{plotErrors}\NormalTok{(errF, }\AttributeTok{nominalQ =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning: Transformation introduced infinite values in continuous y-axis
\end{verbatim}

\begin{verbatim}
## Warning: Transformation introduced infinite values in continuous y-axis
\end{verbatim}

\includegraphics{output_paired/16Sreport_dada2_pair2021-04-30_files/figure-latex/Learn_errors-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Save to file}
\FunctionTok{ggsave}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(readQC.folder, }\StringTok{"/Error\_rates\_per\_sample\_FWD.pdf"}\NormalTok{), }
    \FunctionTok{plotErrors}\NormalTok{(errF, }\AttributeTok{nominalQ =} \ConstantTok{TRUE}\NormalTok{), }\AttributeTok{device =} \StringTok{"pdf"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Saving 10 x 4.5 in image
\end{verbatim}

\begin{verbatim}
## Warning: Transformation introduced infinite values in continuous y-axis

## Warning: Transformation introduced infinite values in continuous y-axis
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Reverse}
\FunctionTok{plotErrors}\NormalTok{(errR, }\AttributeTok{nominalQ =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning: Transformation introduced infinite values in continuous y-axis

## Warning: Transformation introduced infinite values in continuous y-axis
\end{verbatim}

\includegraphics{output_paired/16Sreport_dada2_pair2021-04-30_files/figure-latex/Learn_errors-2.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Save to file}
\FunctionTok{ggsave}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(readQC.folder, }\StringTok{"/Error\_rates\_per\_sample\_REV.pdf"}\NormalTok{), }
    \FunctionTok{plotErrors}\NormalTok{(errR, }\AttributeTok{nominalQ =} \ConstantTok{TRUE}\NormalTok{), }\AttributeTok{device =} \StringTok{"pdf"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Saving 10 x 4.5 in image
\end{verbatim}

\begin{verbatim}
## Warning: Transformation introduced infinite values in continuous y-axis

## Warning: Transformation introduced infinite values in continuous y-axis
\end{verbatim}

The error rates for each possible transition (eg. A-\textgreater C,
A-\textgreater G, \ldots) are shown. Points are the observed error rates
for each consensus quality score. The black line shows the estimated
error rates after convergence. The red line shows the error rates
expected under the nominal definition of the Q-value. If the black line
(the estimated rates) fits the observed rates well, and the error rates
drop with increased quality as expected, then everything looks
reasonable and can proceed with confidence.

\hypertarget{infer-sequence-variants}{%
\subsection{Infer Sequence Variants}\label{infer-sequence-variants}}

This step consists of dereplication, sample inference, and merging of
paired reads

Dereplication combines all identical sequencing reads into into ``unique
sequences'' with a corresponding ``abundance'': the number of reads with
that unique sequence. DADA2 retains a summary of the quality information
associated with each unique sequence. The consensus quality profile of a
unique sequence is the average of the positional qualities from the
dereplicated reads. These quality profiles inform the error model of the
subsequent denoising step, significantly increasing DADA2's accuracy.

The sample inference step performs the core sequence-variant inference
algorithm to the dereplicated data.

Spurious sequence variants are further reduced by merging overlapping
reads. The core function here is mergePairs, which depends on the
forward and reverse re.samples being in matching order at the time they
were dereplicated.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Sample inference of dereplicated reads, and merger of}
\CommentTok{\# paired{-}end reads}
\ControlFlowTok{if}\NormalTok{ (}\SpecialCharTok{!}\NormalTok{pool.samples }\SpecialCharTok{==} \StringTok{"TRUE"}\NormalTok{) \{}
\NormalTok{    mergers }\OtherTok{\textless{}{-}} \FunctionTok{vector}\NormalTok{(}\StringTok{"list"}\NormalTok{, }\FunctionTok{length}\NormalTok{(sample.names))}
    \FunctionTok{names}\NormalTok{(mergers) }\OtherTok{\textless{}{-}}\NormalTok{ sample.names}
    \FunctionTok{names}\NormalTok{(filtFs) }\OtherTok{\textless{}{-}}\NormalTok{ sample.names}
    \FunctionTok{names}\NormalTok{(filtRs) }\OtherTok{\textless{}{-}}\NormalTok{ sample.names}
    \ControlFlowTok{for}\NormalTok{ (sam }\ControlFlowTok{in}\NormalTok{ sample.names) \{}
        \FunctionTok{cat}\NormalTok{(}\StringTok{"Processing:"}\NormalTok{, sam, }\StringTok{"}\SpecialCharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{        derepF }\OtherTok{\textless{}{-}} \FunctionTok{derepFastq}\NormalTok{(filtFs[[sam]])}
\NormalTok{        ddF }\OtherTok{\textless{}{-}} \FunctionTok{dada}\NormalTok{(derepF, }\AttributeTok{err =}\NormalTok{ errF, }\AttributeTok{multithread =}\NormalTok{ threads)}
\NormalTok{        derepR }\OtherTok{\textless{}{-}} \FunctionTok{derepFastq}\NormalTok{(filtRs[[sam]])}
\NormalTok{        ddR }\OtherTok{\textless{}{-}} \FunctionTok{dada}\NormalTok{(derepR, }\AttributeTok{err =}\NormalTok{ errR, }\AttributeTok{multithread =}\NormalTok{ threads)}
\NormalTok{        merger }\OtherTok{\textless{}{-}} \FunctionTok{mergePairs}\NormalTok{(ddF, derepF, ddR, derepR)}
\NormalTok{        mergers[[sam]] }\OtherTok{\textless{}{-}}\NormalTok{ merger}
\NormalTok{    \}}
    \FunctionTok{cat}\NormalTok{(pool.samples)}
    \FunctionTok{rm}\NormalTok{(derepF)}
    \FunctionTok{rm}\NormalTok{(derepR)}
\NormalTok{\} }\ControlFlowTok{else}\NormalTok{ \{}
\NormalTok{    derepFs }\OtherTok{\textless{}{-}} \FunctionTok{derepFastq}\NormalTok{(filtFs, }\AttributeTok{verbose =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{    derepRs }\OtherTok{\textless{}{-}} \FunctionTok{derepFastq}\NormalTok{(filtRs, }\AttributeTok{verbose =} \ConstantTok{TRUE}\NormalTok{)}
    \FunctionTok{names}\NormalTok{(derepFs) }\OtherTok{\textless{}{-}}\NormalTok{ sample.names}
    \FunctionTok{names}\NormalTok{(derepRs) }\OtherTok{\textless{}{-}}\NormalTok{ sample.names}
\NormalTok{    dadaFs }\OtherTok{\textless{}{-}} \FunctionTok{dada}\NormalTok{(derepFs, }\AttributeTok{err =}\NormalTok{ errF, }\AttributeTok{multithread =}\NormalTok{ threads, }
        \AttributeTok{pool =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{    dadaRs }\OtherTok{\textless{}{-}} \FunctionTok{dada}\NormalTok{(derepRs, }\AttributeTok{err =}\NormalTok{ errR, }\AttributeTok{multithread =}\NormalTok{ threads, }
        \AttributeTok{pool =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{    mergers }\OtherTok{\textless{}{-}} \FunctionTok{mergePairs}\NormalTok{(dadaFs, derepFs, dadaRs, derepRs, }\AttributeTok{verbose =} \ConstantTok{TRUE}\NormalTok{)}
    \FunctionTok{cat}\NormalTok{(pool.samples)}
    \FunctionTok{rm}\NormalTok{(derepFs)}
    \FunctionTok{rm}\NormalTok{(derepRs)}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0023_BI_F_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 10285 unique sequences from 53296 total sequences read.
\end{verbatim}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0056_BI_F_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 6843 unique sequences from 47545 total sequences read.
\end{verbatim}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0131_BI_F_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 7450 unique sequences from 46797 total sequences read.
\end{verbatim}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0153_BI_F_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 7331 unique sequences from 39510 total sequences read.
\end{verbatim}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0215_BI_F_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 7861 unique sequences from 40256 total sequences read.
\end{verbatim}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0353_BI_F_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 7184 unique sequences from 49323 total sequences read.
\end{verbatim}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0023_BI_R_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 15600 unique sequences from 53296 total sequences read.
\end{verbatim}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0056_BI_R_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 11149 unique sequences from 47545 total sequences read.
\end{verbatim}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0131_BI_R_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 12109 unique sequences from 46797 total sequences read.
\end{verbatim}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0153_BI_R_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 11085 unique sequences from 39510 total sequences read.
\end{verbatim}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0215_BI_R_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 11940 unique sequences from 40256 total sequences read.
\end{verbatim}

\begin{verbatim}
## Dereplicating sequence entries in Fastq file: /Users/jianshuzhao/Github/dada2_wrapper/scripts/output_paired/filtered_input/Orwoll_BI0353_BI_R_filt.fastq.gz
\end{verbatim}

\begin{verbatim}
## Encountered 12168 unique sequences from 49323 total sequences read.
\end{verbatim}

\begin{verbatim}
## 6 samples were pooled: 276727 reads in 40449 unique sequences.
## 6 samples were pooled: 276727 reads in 66548 unique sequences.
\end{verbatim}

\begin{verbatim}
## 49613 paired-reads (in 704 unique pairings) successfully merged out of 51976 (in 1545 pairings) input.
\end{verbatim}

\begin{verbatim}
## 45841 paired-reads (in 515 unique pairings) successfully merged out of 47042 (in 1038 pairings) input.
\end{verbatim}

\begin{verbatim}
## 44226 paired-reads (in 578 unique pairings) successfully merged out of 46115 (in 1279 pairings) input.
\end{verbatim}

\begin{verbatim}
## 37388 paired-reads (in 667 unique pairings) successfully merged out of 38980 (in 1328 pairings) input.
\end{verbatim}

\begin{verbatim}
## 37812 paired-reads (in 643 unique pairings) successfully merged out of 39519 (in 1367 pairings) input.
\end{verbatim}

\begin{verbatim}
## 46633 paired-reads (in 463 unique pairings) successfully merged out of 48652 (in 1114 pairings) input.
\end{verbatim}

\begin{verbatim}
## TRUE
\end{verbatim}

\hypertarget{construct-sequence-table}{%
\subsection{Construct sequence table}\label{construct-sequence-table}}

We can now construct a ``sequence table'' of our samples, a
higher-resolution version of the ``OTU table'' produced by classical
methods:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{seqtab }\OtherTok{\textless{}{-}} \FunctionTok{makeSequenceTable}\NormalTok{(mergers)}
\FunctionTok{dim}\NormalTok{(seqtab)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]    6 1452
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Inspect distribution of sequence lengths}
\FunctionTok{table}\NormalTok{(}\FunctionTok{nchar}\NormalTok{(}\FunctionTok{getSequences}\NormalTok{(seqtab)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
##  252  253  254 
##   87 1357    8
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# The sequence table is a matrix with rows corresponding to}
\CommentTok{\# (and named by) the samples, and columns corresponding to}
\CommentTok{\# (and named by) the sequence variants.}
\end{Highlighting}
\end{Shaded}

\hypertarget{remove-chimeras}{%
\subsection{Remove chimeras}\label{remove-chimeras}}

The core dada method removes substitution and indel errors, but chimeras
remain. Fortunately, the accuracy of the sequences after denoising makes
identifying chimeras simpler than it is when dealing with fuzzy OTUs:
all sequences which can be exactly reconstructed as a bimera (two-parent
chimera) from more abundant sequences.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Remove chimeric sequences:}
\NormalTok{seqtab.nochim }\OtherTok{\textless{}{-}} \FunctionTok{removeBimeraDenovo}\NormalTok{(seqtab, }\AttributeTok{method =} \StringTok{"consensus"}\NormalTok{, }
    \AttributeTok{multithread =}\NormalTok{ threads, }\AttributeTok{verbose =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Identified 934 bimeras out of 1452 input sequences.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{dim}\NormalTok{(seqtab.nochim)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]   6 518
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# ratio of chimeric sequence reads}
\DecValTok{1} \SpecialCharTok{{-}} \FunctionTok{sum}\NormalTok{(seqtab.nochim)}\SpecialCharTok{/}\FunctionTok{sum}\NormalTok{(seqtab)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.1049661
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# write sequence variants count table to file}
\FunctionTok{write.table}\NormalTok{(}\FunctionTok{t}\NormalTok{(seqtab.nochim), }\FunctionTok{paste0}\NormalTok{(output.dir, }\StringTok{"/all\_samples\_SV{-}counts.tsv"}\NormalTok{), }
    \AttributeTok{sep =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{, }\AttributeTok{eol =} \StringTok{"}\SpecialCharTok{\textbackslash{}n}\StringTok{"}\NormalTok{, }\AttributeTok{quote =}\NormalTok{ F, }\AttributeTok{col.names =} \ConstantTok{NA}\NormalTok{)}
\CommentTok{\# write OTU table to file}
\FunctionTok{saveRDS}\NormalTok{(seqtab.nochim, }\FunctionTok{paste0}\NormalTok{(output.dir, }\StringTok{"/seqtab\_final.rds"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\textbf{IMPORTANT:} Most of your \textbf{reads} should remain after
chimera removal (it is not uncommon for a majority of \textbf{sequence
variants} to be removed though). If most of your reads were removed as
chimeric, upstream processing may need to be revisited. In almost all
cases this is caused by primer sequences with ambiguous nucleotides that
were not removed prior to beginning the DADA2 pipeline.

\hypertarget{track-reads-through-the-pipeline}{%
\subsection{Track reads through the
pipeline}\label{track-reads-through-the-pipeline}}

As a final check of the progress, look at the number of reads that made
it through each step in the pipeline. This is a great place to do a last
sanity check. Outside of filtering (depending on how stringent you want
to be) there should no step in which a majority of reads are lost. If a
majority of reads failed to merge, you may need to revisit the truncLen
parameter used in the filtering step and make sure that the truncated
reads span your amplicon. If a majority of reads failed to pass the
chimera check, you may need to revisit the removal of primers, as the
ambiguous nucleotides in unremoved primers interfere with chimera
identification.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{getN }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(x) }\FunctionTok{sum}\NormalTok{(}\FunctionTok{getUniques}\NormalTok{(x))}
\NormalTok{track }\OtherTok{\textless{}{-}} \FunctionTok{cbind}\NormalTok{(rd.counts, }\FunctionTok{sapply}\NormalTok{(mergers, getN), }\FunctionTok{rowSums}\NormalTok{(seqtab), }
    \FunctionTok{rowSums}\NormalTok{(seqtab.nochim))}
\FunctionTok{colnames}\NormalTok{(track) }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"input"}\NormalTok{, }\StringTok{"filtered"}\NormalTok{, }\StringTok{"ratio"}\NormalTok{, }\StringTok{"merged"}\NormalTok{, }
    \StringTok{"tabled"}\NormalTok{, }\StringTok{"nonchim"}\NormalTok{)}
\FunctionTok{rownames}\NormalTok{(track) }\OtherTok{\textless{}{-}}\NormalTok{ sample.names}
\CommentTok{\# print table}
\NormalTok{track}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                  input filtered ratio merged tabled nonchim
## Orwoll_BI0023_BI 62724    53296  0.85  49613  49613   45189
## Orwoll_BI0056_BI 55342    47545  0.86  45841  45841   39776
## Orwoll_BI0131_BI 55144    46797  0.85  44226  44226   42032
## Orwoll_BI0153_BI 44610    39510  0.89  37388  37388   33152
## Orwoll_BI0215_BI 48227    40256  0.83  37812  37812   33457
## Orwoll_BI0353_BI 54271    49323  0.91  46633  46633   40457
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# save to file}
\FunctionTok{write.table}\NormalTok{(track, }\FunctionTok{paste0}\NormalTok{(readQC.folder, }\StringTok{"/Read\_counts\_at\_each\_step.tsv"}\NormalTok{), }
    \AttributeTok{sep =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{, }\AttributeTok{quote =}\NormalTok{ F, }\AttributeTok{eol =} \StringTok{"}\SpecialCharTok{\textbackslash{}n}\StringTok{"}\NormalTok{, }\AttributeTok{col.names =} \ConstantTok{NA}\NormalTok{)}

\CommentTok{\# rename fasta header and write to file}
\NormalTok{seqtab.nochim\_trans }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{t}\NormalTok{(seqtab.nochim)) }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{rownames\_to\_column}\NormalTok{(}\AttributeTok{var =} \StringTok{"sequence"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{rowid\_to\_column}\NormalTok{(}\AttributeTok{var =} \StringTok{"OTUNumber"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{mutate}\NormalTok{(}\AttributeTok{OTUNumber =} \FunctionTok{sprintf}\NormalTok{(}\StringTok{"otu\%04d"}\NormalTok{, OTUNumber)) }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{mutate}\NormalTok{(}\AttributeTok{sequence =} \FunctionTok{str\_replace\_all}\NormalTok{(sequence, }\StringTok{"({-}|}\SpecialCharTok{\textbackslash{}\textbackslash{}}\StringTok{.)"}\NormalTok{, }\StringTok{""}\NormalTok{))}

\NormalTok{df }\OtherTok{\textless{}{-}}\NormalTok{ seqtab.nochim\_trans}
\NormalTok{seq\_out }\OtherTok{\textless{}{-}}\NormalTok{ Biostrings}\SpecialCharTok{::}\FunctionTok{DNAStringSet}\NormalTok{(df}\SpecialCharTok{$}\NormalTok{sequence)}

\FunctionTok{names}\NormalTok{(seq\_out) }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{OTUNumber}

\NormalTok{Biostrings}\SpecialCharTok{::}\FunctionTok{writeXStringSet}\NormalTok{(seq\_out, }\FunctionTok{str\_c}\NormalTok{(output.dir, }\StringTok{"/ASV\_rep\_no\_tax.fasta"}\NormalTok{), }
    \AttributeTok{compress =} \ConstantTok{FALSE}\NormalTok{, }\AttributeTok{width =} \DecValTok{20000}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{align-sequences-and-reconstruct-phylogeny}{%
\subsection{Align sequences and reconstruct
phylogeny}\label{align-sequences-and-reconstruct-phylogeny}}

Multiple sequence alignment of resolved sequence variants is used to
generate a phylogenetic tree, which is required for calculating UniFrac
beta-diversity distances between microbiome samples.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Get sequences}
\NormalTok{seqs }\OtherTok{\textless{}{-}} \FunctionTok{getSequences}\NormalTok{(seqtab.nochim)}
\CommentTok{\# save non{-}chimera, representative sequences}


\FunctionTok{names}\NormalTok{(seqs) }\OtherTok{\textless{}{-}}\NormalTok{ seqs  }\CommentTok{\# This propagates to the tip labels of the tree}
\CommentTok{\# Multiple seqeuence alignment}
\NormalTok{mult }\OtherTok{\textless{}{-}} \FunctionTok{msa}\NormalTok{(seqs, }\AttributeTok{method =} \StringTok{"ClustalOmega"}\NormalTok{, }\AttributeTok{type =} \StringTok{"dna"}\NormalTok{, }\AttributeTok{order =} \StringTok{"input"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## using Gonnet
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Save msa to file; convert first to phangorn object}
\NormalTok{phang.align }\OtherTok{\textless{}{-}} \FunctionTok{as.phyDat}\NormalTok{(mult, }\AttributeTok{type =} \StringTok{"DNA"}\NormalTok{, }\AttributeTok{names =} \FunctionTok{getSequences}\NormalTok{(seqtab.nochim))}
\FunctionTok{write.phyDat}\NormalTok{(phang.align, }\AttributeTok{format =} \StringTok{"fasta"}\NormalTok{, }\AttributeTok{file =} \FunctionTok{paste0}\NormalTok{(output.dir, }
    \StringTok{"/msa.fasta"}\NormalTok{))}

\CommentTok{\# Call FastTree (via \textquotesingle{}system\textquotesingle{}) to reconstruct phylogeny}
\ControlFlowTok{if}\NormalTok{ (}\FunctionTok{unname}\NormalTok{(}\FunctionTok{Sys.info}\NormalTok{()[}\StringTok{"sysname"}\NormalTok{]) }\SpecialCharTok{==} \StringTok{"Linux"}\NormalTok{) \{}
    \FunctionTok{system}\NormalTok{(}\FunctionTok{paste}\NormalTok{(}\StringTok{"../dependencies/FastTreeMP\_Linux {-}gtr {-}nt "}\NormalTok{, }
\NormalTok{        output.dir, }\StringTok{"/msa.fasta \textgreater{} "}\NormalTok{, output.dir, }\StringTok{"/FastTree.tre"}\NormalTok{, }
        \AttributeTok{sep =} \StringTok{""}\NormalTok{))}
\NormalTok{\} }\ControlFlowTok{else}\NormalTok{ \{}
    \FunctionTok{system}\NormalTok{(}\FunctionTok{paste}\NormalTok{(}\StringTok{"../dependencies/FastTreeMP\_Darwin {-}gtr {-}nt "}\NormalTok{, }
\NormalTok{        output.dir, }\StringTok{"/msa.fasta \textgreater{} "}\NormalTok{, output.dir, }\StringTok{"/FastTree.tre"}\NormalTok{, }
        \AttributeTok{sep =} \StringTok{""}\NormalTok{))}
\NormalTok{\}}

\FunctionTok{detach}\NormalTok{(}\StringTok{"package:phangorn"}\NormalTok{, }\AttributeTok{unload =} \ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{detach}\NormalTok{(}\StringTok{"package:msa"}\NormalTok{, }\AttributeTok{unload =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{assign-taxonomy}{%
\subsection{Assign taxonomy}\label{assign-taxonomy}}

The assignTaxonomy function takes a set of sequences and a training set
of taxonomically classified sequences, and outputs the taxonomic
assignments with at least minBoot bootstrap confidence. Formatted
training datasets for taxonomic assignments can be downloaded from here
\url{https://benjjneb.github.io/dada2/training.html}.

\texttt{assignTaxonomy(\ ...\ )} implements the RDP naive Bayesian
classifier method described in Wang et al.~2007. In short, the kmer
profile of the sequences to be classified are compared against the kmer
profiles of all sequences in a training set of sequences with assigned
taxonomies. The reference sequence with the most similar profile is used
to assign taxonomy to the query sequence, and then a bootstrapping
approach is used to assess the confidence assignment at each taxonomic
level. The return value of assignTaxonomy(\ldots) is a character matrix,
with each row corresponding to an input sequence, and each column
corresponding to a taxonomic level.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Assign taxonomy: print(paste(\textquotesingle{}Assign taxonomy (Greengene)}
\CommentTok{\# starts at\textquotesingle{},Sys.time, sep = \textquotesingle{} \textquotesingle{}))}
\NormalTok{taxa.gg13\_8 }\OtherTok{\textless{}{-}} \FunctionTok{assignTaxonomy}\NormalTok{(seqtab.nochim, }\StringTok{"../reference\_dbs\_16S/gg\_13\_8\_train\_set\_97.fa.gz"}\NormalTok{, }
    \AttributeTok{multithread =}\NormalTok{ threads, }\AttributeTok{tryRC =} \ConstantTok{TRUE}\NormalTok{)}

\CommentTok{\# Print first 6 rows of taxonomic assignment}
\FunctionTok{unname}\NormalTok{(}\FunctionTok{head}\NormalTok{(taxa.gg13\_8))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      [,1]          [,2]                [,3]                    
## [1,] "k__Bacteria" "p__Bacteroidetes"  "c__Bacteroidia"        
## [2,] "k__Bacteria" "p__Proteobacteria" "c__Gammaproteobacteria"
## [3,] "k__Bacteria" "p__Bacteroidetes"  "c__Bacteroidia"        
## [4,] "k__Bacteria" "p__Firmicutes"     "c__Clostridia"         
## [5,] "k__Bacteria" "p__Bacteroidetes"  "c__Bacteroidia"        
## [6,] "k__Bacteria" "p__Bacteroidetes"  "c__Bacteroidia"        
##      [,4]                   [,5]                    [,6]                 
## [1,] "o__Bacteroidales"     "f__Bacteroidaceae"     "g__Bacteroides"     
## [2,] "o__Enterobacteriales" "f__Enterobacteriaceae" "g__Klebsiella"      
## [3,] "o__Bacteroidales"     "f__Bacteroidaceae"     "g__Bacteroides"     
## [4,] "o__Clostridiales"     "f__Ruminococcaceae"    "g__Faecalibacterium"
## [5,] "o__Bacteroidales"     "f__Bacteroidaceae"     "g__Bacteroides"     
## [6,] "o__Bacteroidales"     "f__Bacteroidaceae"     "g__Bacteroides"     
##      [,7]            
## [1,] "s__"           
## [2,] "s__"           
## [3,] "s__caccae"     
## [4,] "s__prausnitzii"
## [5,] "s__ovatus"     
## [6,] "s__ovatus"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Replace NAs in taxonomy assignment table with prefix}
\CommentTok{\# corresponding to tax rank}
\NormalTok{taxa.gg13\_8}\FloatTok{.2} \OtherTok{\textless{}{-}} \FunctionTok{replaceNA.in.assignedTaxonomy}\NormalTok{(taxa.gg13\_8)}

\CommentTok{\# Write taxa table to file}
\FunctionTok{write.table}\NormalTok{(taxa.gg13\_8}\FloatTok{.2}\NormalTok{, }\FunctionTok{paste0}\NormalTok{(output.dir, }\StringTok{"/all\_samples\_GG13{-}8{-}taxonomy.tsv"}\NormalTok{), }
    \AttributeTok{sep =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{, }\AttributeTok{eol =} \StringTok{"}\SpecialCharTok{\textbackslash{}n}\StringTok{"}\NormalTok{, }\AttributeTok{quote =}\NormalTok{ F, }\AttributeTok{col.names =} \ConstantTok{NA}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{merge-otu-and-gg13-8-taxonomy-tables}{%
\subsection{Merge OTU and GG13-8 taxonomy
tables}\label{merge-otu-and-gg13-8-taxonomy-tables}}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{otu.gg.tax.table }\OtherTok{\textless{}{-}} \FunctionTok{merge}\NormalTok{(}\FunctionTok{t}\NormalTok{(seqtab.nochim), taxa.gg13\_8}\FloatTok{.2}\NormalTok{, }\AttributeTok{by =} \StringTok{"row.names"}\NormalTok{)}
\FunctionTok{rownames}\NormalTok{(otu.gg.tax.table) }\OtherTok{\textless{}{-}}\NormalTok{ otu.gg.tax.table[, }\DecValTok{1}\NormalTok{]}
\NormalTok{otu.gg.tax.table }\OtherTok{\textless{}{-}}\NormalTok{ otu.gg.tax.table[, }\SpecialCharTok{{-}}\DecValTok{1}\NormalTok{]}

\FunctionTok{write.table}\NormalTok{(otu.gg.tax.table, }\FunctionTok{paste0}\NormalTok{(output.dir, }\StringTok{"/all\_samples\_SV{-}counts\_and\_GG13{-}8{-}taxonomy.tsv"}\NormalTok{), }
    \AttributeTok{sep =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{, }\AttributeTok{eol =} \StringTok{"}\SpecialCharTok{\textbackslash{}n}\StringTok{"}\NormalTok{, }\AttributeTok{quote =}\NormalTok{ F, }\AttributeTok{col.names =} \ConstantTok{NA}\NormalTok{)}
\DocumentationTok{\#\# print(paste(\textquotesingle{}Assign taxonomy (Greengene) finishes}
\DocumentationTok{\#\# at\textquotesingle{},Sys.time, sep = \textquotesingle{} \textquotesingle{}))}
\end{Highlighting}
\end{Shaded}

For RDP and Silva, taxonomic assignment to species level is a two-step
process. Fast and appropriate species-level assignment from 16S data is
provided by the \texttt{assignSpecies(\ ...\ )} method.
\texttt{assignSpecies(\ ...\ )} uses exact string matching against a
reference database to assign Genus species binomials. In short, query
sequence are compared against all reference sequences that had binomial
genus-species nomenclature assigned, and the genus-species of all exact
matches are recorded and returned if it is unambiguous.

The convenience function \texttt{addSpecies(\ ...\ )} takes as input a
taxonomy table, and outputs a table with an added species column. Only
those genus-species binomials which are consistent with the genus
assigned in the provided taxonomy table are retained in the output. See
here for more on taxonomic assignment
\url{https://benjjneb.github.io/dada2/assign.html}.

\hypertarget{asign-silva-and-rdp-taxonomies-and-merge-with-otu-table}{%
\subsection{Asign SILVA and RDP taxonomies and merge with OTU
table}\label{asign-silva-and-rdp-taxonomies-and-merge-with-otu-table}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Assign SILVA taxonomy print(paste(\textquotesingle{}Assign taxonomy (Silva)}
\CommentTok{\# starts at\textquotesingle{},Sys.time, sep = \textquotesingle{} \textquotesingle{}))}
\NormalTok{taxa.silva }\OtherTok{\textless{}{-}} \FunctionTok{assignTaxonomy}\NormalTok{(seqtab.nochim, }\StringTok{"../reference\_dbs\_16S/silva\_nr\_v128\_train\_set.fa.gz"}\NormalTok{, }
    \AttributeTok{multithread =}\NormalTok{ threads)}
\FunctionTok{saveRDS}\NormalTok{(taxa.silva, }\FunctionTok{paste0}\NormalTok{(output.dir, }\StringTok{"/taxa.silva.1.rds"}\NormalTok{))}
\CommentTok{\# Replace NAs in taxonomy assignment table with prefix}
\CommentTok{\# corresponding to tax rank}
\NormalTok{taxa.silva}\FloatTok{.2} \OtherTok{\textless{}{-}} \FunctionTok{replaceNA.in.assignedTaxonomy}\NormalTok{(taxa.silva)}
\FunctionTok{saveRDS}\NormalTok{(taxa.silva}\FloatTok{.2}\NormalTok{, }\FunctionTok{paste0}\NormalTok{(output.dir, }\StringTok{"/taxa.silva.2.rds"}\NormalTok{))}
\CommentTok{\# OMIT APPENDING SPECIES FOR SILVA DUE TO MEMORY CONSTRAINTS}
\CommentTok{\# Append species. Note that appending the argument}
\CommentTok{\# \textquotesingle{}allowMultiple=3\textquotesingle{} will return up to 3 different matched}
\CommentTok{\# species, but if 4 or more are matched it returns NA.}
\CommentTok{\# taxa.silva.species \textless{}{-} addSpecies(taxa.silva,}
\CommentTok{\# \textquotesingle{}/n/huttenhower\_lab/data/dada2\_reference\_databases/silva\_species\_assignment\_v128.fa.gz\textquotesingle{})}

\CommentTok{\# Merge with OTU table and save to file}
\NormalTok{otu.silva.tax.table }\OtherTok{\textless{}{-}} \FunctionTok{merge}\NormalTok{(}\FunctionTok{t}\NormalTok{(seqtab.nochim), taxa.silva}\FloatTok{.2}\NormalTok{, }
    \AttributeTok{by =} \StringTok{"row.names"}\NormalTok{)}
\FunctionTok{rownames}\NormalTok{(otu.silva.tax.table) }\OtherTok{\textless{}{-}}\NormalTok{ otu.silva.tax.table[, }\DecValTok{1}\NormalTok{]}
\NormalTok{otu.silva.tax.table }\OtherTok{\textless{}{-}}\NormalTok{ otu.silva.tax.table[, }\SpecialCharTok{{-}}\DecValTok{1}\NormalTok{]}


\FunctionTok{write.table}\NormalTok{(otu.silva.tax.table, }\FunctionTok{paste0}\NormalTok{(output.dir, }\StringTok{"/all\_samples\_SV{-}counts\_and\_SILVA{-}taxonomy.tsv"}\NormalTok{), }
    \AttributeTok{sep =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{, }\AttributeTok{eol =} \StringTok{"}\SpecialCharTok{\textbackslash{}n}\StringTok{"}\NormalTok{, }\AttributeTok{quote =}\NormalTok{ F, }\AttributeTok{col.names =} \ConstantTok{NA}\NormalTok{)}
\DocumentationTok{\#\# print(paste(\textquotesingle{}Assign taxonomy (Silva) finishes at\textquotesingle{},Sys.time,}
\DocumentationTok{\#\# sep = \textquotesingle{} \textquotesingle{})) Assign RDP taxonomy print(paste(\textquotesingle{}Assign}
\DocumentationTok{\#\# taxonomy (RDP) starts at\textquotesingle{},Sys.time, sep = \textquotesingle{} \textquotesingle{}))}
\NormalTok{taxa.rdp }\OtherTok{\textless{}{-}} \FunctionTok{assignTaxonomy}\NormalTok{(seqtab.nochim, }\StringTok{"../reference\_dbs\_16S/rdp\_train\_set\_18.fa.gz"}\NormalTok{, }
    \AttributeTok{multithread =}\NormalTok{ threads)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in .Call2("fasta_index", filexp_list, nrec, skip, seek.first.rec, :
## reading FASTA file ../reference_dbs_16S/rdp_train_set_18.fa.gz: ignored 9
## invalid one-letter sequence codes
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Replace NAs in taxonomy assignment table with prefix}
\CommentTok{\# corresponding to tax rank}
\NormalTok{taxa.rdp}\FloatTok{.2} \OtherTok{\textless{}{-}} \FunctionTok{replaceNA.in.assignedTaxonomy}\NormalTok{(taxa.rdp)}

\CommentTok{\# OMIT APPENDING SPECIES FOR RDP DUE TO MEMORY CONSTRAINTS}
\CommentTok{\# Append species. Note that appending the argument}
\CommentTok{\# \textquotesingle{}allowMultiple=3\textquotesingle{} will return up to 3 different matched}
\CommentTok{\# species, but if 4 or more are matched it returns NA.}
\CommentTok{\# taxa.rdp.species \textless{}{-} addSpecies(taxa.rdp,}
\CommentTok{\# \textquotesingle{}/n/huttenhower\_lab/data/dada2\_reference\_databases/rdp\_species\_assignment\_16.fa.gz\textquotesingle{})}


\CommentTok{\# Merge with OTU table and save to file}
\NormalTok{otu.rdp.tax.table }\OtherTok{\textless{}{-}} \FunctionTok{merge}\NormalTok{(}\FunctionTok{t}\NormalTok{(seqtab.nochim), taxa.rdp}\FloatTok{.2}\NormalTok{, }\AttributeTok{by =} \StringTok{"row.names"}\NormalTok{)}
\FunctionTok{rownames}\NormalTok{(otu.rdp.tax.table) }\OtherTok{\textless{}{-}}\NormalTok{ otu.rdp.tax.table[, }\DecValTok{1}\NormalTok{]}
\NormalTok{otu.rdp.tax.table }\OtherTok{\textless{}{-}}\NormalTok{ otu.rdp.tax.table[, }\SpecialCharTok{{-}}\DecValTok{1}\NormalTok{]}
\CommentTok{\# extract representative sequences and save to file}

\FunctionTok{write.table}\NormalTok{(otu.rdp.tax.table, }\FunctionTok{paste0}\NormalTok{(output.dir, }\StringTok{"/all\_samples\_SV{-}counts\_and\_RDP{-}taxonomy.tsv"}\NormalTok{), }
    \AttributeTok{sep =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{, }\AttributeTok{eol =} \StringTok{"}\SpecialCharTok{\textbackslash{}n}\StringTok{"}\NormalTok{, }\AttributeTok{quote =}\NormalTok{ F, }\AttributeTok{col.names =} \ConstantTok{NA}\NormalTok{)}
\DocumentationTok{\#\# print(paste(\textquotesingle{}Assign taxonomy (RDP) finishes at\textquotesingle{},Sys.time,}
\DocumentationTok{\#\# sep = \textquotesingle{} \textquotesingle{}))}
\end{Highlighting}
\end{Shaded}

\newpage

Session Info:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{sessionInfo}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## R version 4.0.3 (2020-10-10)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## BLAS:   /Users/jianshuzhao/miniconda3/lib/libblas.3.9.0.dylib
## LAPACK: /Users/jianshuzhao/miniconda3/lib/liblapack.3.9.0.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] stringr_1.4.0               RcppParallel_5.1.2         
##  [3] gridExtra_2.3               ape_5.5                    
##  [5] dada2_1.18.0                Rcpp_1.0.6                 
##  [7] Rsamtools_2.6.0             Biostrings_2.58.0          
##  [9] XVector_0.30.0              BiocParallel_1.24.1        
## [11] SummarizedExperiment_1.20.0 Biobase_2.50.0             
## [13] MatrixGenerics_1.2.1        matrixStats_0.58.0         
## [15] GenomicRanges_1.42.0        GenomeInfoDb_1.26.7        
## [17] IRanges_2.24.1              S4Vectors_0.28.1           
## [19] BiocGenerics_0.36.1         cowplot_1.1.1              
## [21] RCurl_1.98-1.3              devtools_2.4.0             
## [23] usethis_2.0.1               knitr_1.31                 
## [25] rmarkdown_2.7               tibble_3.1.1               
## [27] ggplot2_3.3.3               mgcv_1.8-34                
## [29] nlme_3.1-152                dplyr_1.0.5                
## [31] BiocManager_1.30.12        
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-7             fs_1.5.0                 RColorBrewer_1.1-2      
##  [4] rprojroot_2.0.2          tools_4.0.3              utf8_1.2.1              
##  [7] R6_2.5.0                 DBI_1.1.1                colorspace_2.0-0        
## [10] withr_2.4.2              tidyselect_1.1.1         prettyunits_1.1.1       
## [13] processx_3.5.1           compiler_4.0.3           cli_2.5.0               
## [16] formatR_1.9              desc_1.3.0               DelayedArray_0.16.3     
## [19] labeling_0.4.2           scales_1.1.1             quadprog_1.5-8          
## [22] callr_3.7.0              digest_0.6.27            jpeg_0.1-8.1            
## [25] pkgconfig_2.0.3          htmltools_0.5.1.1        sessioninfo_1.1.1       
## [28] highr_0.9                fastmap_1.1.0            rlang_0.4.10            
## [31] rstudioapi_0.13          farver_2.1.0             generics_0.1.0          
## [34] hwriter_1.3.2            magrittr_2.0.1           GenomeInfoDbData_1.2.4  
## [37] Matrix_1.3-2             munsell_0.5.0            fansi_0.4.2             
## [40] lifecycle_1.0.0          yaml_2.2.1               stringi_1.5.3           
## [43] zlibbioc_1.36.0          pkgbuild_1.2.0           plyr_1.8.6              
## [46] grid_4.0.3               crayon_1.4.1             lattice_0.20-41         
## [49] splines_4.0.3            ps_1.6.0                 pillar_1.6.0            
## [52] igraph_1.2.6             codetools_0.2-18         reshape2_1.4.4          
## [55] pkgload_1.2.1            fastmatch_1.1-0          glue_1.4.2              
## [58] evaluate_0.14            ShortRead_1.48.0         latticeExtra_0.6-29     
## [61] remotes_2.3.0            vctrs_0.3.8              png_0.1-7               
## [64] testthat_3.0.2           gtable_0.3.0             purrr_0.3.4             
## [67] assertthat_0.2.1         cachem_1.0.4             xfun_0.20               
## [70] GenomicAlignments_1.26.0 memoise_2.0.0            ellipsis_0.3.2
\end{verbatim}

\end{document}
