---
title: "Diversity analysis of 16S amplicon after DADA2"
author: "Jianshu Zhao"
date: "`r Sys.Date()`"
output: 
  pdf_document: default
---

<br>

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width = 10)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
```


<br>

## Introduction
Diversity analysis for profiling microbial community based on DADA2. Alpha diversity metric include Chao1, Shannon, Simpson and beta diversity metric include Bray-Curits, UniFrac and Jaccard distance will be calculated
<br>

## Load R packages:  
```{r load_packages, echo=FALSE}

# load packages
library(ggplot2) #Package to generate plots
library(vegan) #community ecology package
library(scales) #To modify scale of plot's axis
library(metagMisc) #Package to manipulate phyloseq objects
library(dplyr) #package used for data manipulation 
library(reshape2) #package used for data manipulation 
library(data.table) #package used for data manipulation 
library(vegan)
library(iNEXT)
library(ape)
library(GUniFrac)
library(phyloseq); packageVersion("phyloseq")
```

<br>
## Packages used
Phyloseq package will be used to import, store and analyze, and graphically display complex
#phylogenetic sequencing data that has already been clustered into Operational Taxonomic Units (OTUs),
# -or in our case ASVs- especially when there is associated sample data, 
#phylogenetic tree, and/or taxonomic assignment of the ASVs. 
#This package leverages many of the tools available in R for ecology and phylogenetic analysis
#(vegan, ade4, ape, picante), while also using advanced/flexible graphic systems (ggplot2) to easily
#produce publication-quality graphics of complex phylogenetic data. phyloseq uses a specialized system
#of S4 classes to store all related phylogenetic sequencing data as single experiment-level object,
#making it easier to share data and reproduce analyses." 
<br>


```{r Reading community matrix and calculate diversity metrics}
setwd(output.dir)
seqtab.nochim <- readRDS("seqtab_final.rds")
tre <- read.tre(Fasttree.tre)



```


```{r generating figures}
a= ggplot()
```
